### PR TITLE
Added FriendlyError class, some lexer errors.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Created an abstract class called FriendlyError (tm), which has a reusable template toString method.  Ideally, we could write very clear and concise error messages.  It has a single subclass currently called LexerError.  This gets thrown occurs when there is a double < or > like '<<div' or 'div>>'.  To support these error formats, I changed the implementation of the StringScanner to a LineScanner so we can have column information.

screen shot 2016-09-30 at 6 03 11 pm

also I wrote a test!

### CODE REVIEW

**Comment by jonahwilliams** [comment]
I misspelled unnecessary >.>

File: lib/src/error.dart [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import 'package:source_span/source_span.dart';
+
+/// A pretty error format for the project.
+abstract class FriendlyError extends Error {
+  /// the line number of the error.
+  String get line;
```
Comment (matanlurey) [inline lib/src/error.dart]: Can most of this just be a `SourceSpan` object, i.e.

``` dart
abstract class SourceError {
  String get message;
  SourceSpan get source;
}
```

File: lib/src/error.dart [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  /// The source span where the error was encountered, from column 0.
+  final SourceSpan span;
+
+  /// Additional lexer context into the failure.
+  final LexerErrorKind kind;
+
+  /// Used to signal possible malformed html to the parser.
+  LexerError(this.span, this.kind);
+
+  @override
```
Comment (matanlurey) [inline lib/src/error.dart:34]: nit: Newline after each getter

File: lib/src/lexer.dart [hunk 3]
```diff
@@ -116,15 +118,15 @@ enum HtmlLexerState {
 
 /// Produces [HtmlToken]s from a [String].
 class HtmlLexer {
-  final StringScanner _scanner;
+  final LineScanner _scanner;
```
Comment (matanlurey) [inline lib/src/lexer.dart:126]: Nice, didn't realize I could just that!

**Review by matanlurey — CHANGES_REQUESTED** [review]
LGTM, but can you rebase against master?

File: lib/src/error.dart (hunk 1 above)
Comment (jonahwilliams) [reply lib/src/error.dart]: Done

File: lib/src/error.dart (hunk 2 above)
Comment (jonahwilliams) [reply lib/src/error.dart:34]: Done

File: lib/src/lexer.dart (hunk 3 above)
Comment (jonahwilliams) [reply lib/src/lexer.dart:126]: I think I am obligated to comment here

**Review by matanlurey — APPROVED** [review]

**Comment by matanlurey** [comment]
LGTM/Approval, thanks @jonahwilliams. Feel free to squash and merge - or if you can approve #8 and #9 then I don't mind helping you rebase.

**Pull request merged**